cosmic-panel-bin/src/notifications.rs

```diff
@@ -0,0 +1,179 @@
+    info!("Inserting channel into event loop");
+
+    // insert channel for spaces to send pending nortification appplet to
+    let (tx, rx) = calloop::channel::sync_channel(10);
+
+    loop_handle
+        .insert_source(
+            rx,
+            move |msg: calloop::channel::Event<PendingAppletEvent>, _, state| match msg {
+                calloop::channel::Event::Msg(msg) => match msg {
+                    PendingAppletEvent::Add(id, stream) => {
+                        info!("Received pending notification applet for space {}", &id);
+                        if !state.space.notification_applet_spaces.contains(&id) {
+                            // state.space.pending_notification_applet_ids.push(msg);
+                            if let Err(err) = write_socket(&daemon_socket, state, (id, stream)) {
+                                error!(
+                                    "Failed to send pending notification applet to daemon: {}",
+                                    err
+                                );
+                            }
+                        }
+                    }
+                    PendingAppletEvent::Remove(id) => {
+                        info!("Removing pending notification applet for space {}", &id);
+                        state.space.notification_applet_spaces.remove(&id);
+                    }
+                },
+                calloop::channel::Event::Closed => {
+                    warn!("Notification channel closed");
+                }
+            },
+        )
+        .map_err(|err| {
+            anyhow::anyhow!(
+                "Failed to insert notification channel into event loop: {}",
+                err
+            )
+        })?;
+
+    info!("Inserting session socket into event loop");
+    // insert source for daemon socket
+    loop_handle
+        .insert_source(
+            Generic::new(daemon_socket_clone, Interest::BOTH, Mode::Edge),
+            move |interest, stream, state: &mut GlobalState<SpaceContainer>| {
+                if interest.error {
+                    error!("Error on session socket");
+                    return Ok(PostAction::Remove);
+                }
+
+                if interest.readable {
+                    if let Err(err) = read_socket(stream, state) {
+                        error!("Error reading from session socket: {}", err);
+                        return Ok(PostAction::Continue);
+                    }
+                }
+
+                Ok(PostAction::Continue)
+            },
+        )
+        .with_context(|| "Failed to init the cosmic session socket source")?;
+
+    Ok(tx)
+}
+
+fn read_socket(stream: &UnixStream, state: &mut GlobalState<SpaceContainer>) -> Result<()> {
+    // every message is a u32 id, and a socket fd
+    info!("Reading from notification daemon socket");
+    let mut buf = [0u8; 128];
+    let mut fd_buf = [0i32; 32];
+    while let Ok((msg_cnt, fd_cnt)) = stream.recv_with_fd(&mut buf, &mut fd_buf) {
```

Comment:
Currently this is ignoring the specific error returned from `recv_with_fd`. It should probably check that, and return `Ok(())` for `EWOULDBLOCK`/`EAGAIN`, and `Err(..)` otherwise?